### DISCUSSION TITLE
Scanner RPC Requirements

### DISCUSSION THREAD

**Discussion by teor2345** [ZcashFoundation/zebra#8153] [Design]
## Motivation

We want to provide secure access to the shielded blockchain scanner.

## Goals

Based on the discussion in this week's Zebra team sync, we have the following goals for the RPC interface:
- Authenticates connections to make sure they are authorised to add or remove keys, or get results
- Encrypts connections to avoid leaking keys or results
- Requires proof of knowledge of keys before accessing results or deleting that key
- Can be used via Tor

Additional security requirement:
- The client should authenticate the server before sending private viewing keys.

<details>
  <summary><h3>Background & Security Notes</h3></summary>

#### `zcashd` authentication

`zcashd` currently authenticates using either:
- proof of knowledge of the contents of a local cookie file, or
- username and password.

Passwords might be deprecated, we should ask the ECC team why that happened, and if they've made a final decision.

#### Tor compatibility

To be used via Tor, the RPC library must communicate over either:
- a [`PeerTransport: AsyncRead + AsyncWrite + ...`](https://github.com/ZcashFoundation/zebra/blob/704f88eb412c9e5c0f0565c81250c6c50818275c/zebra-network/src/isolated.rs#L55), or
- [`arti::DataStream`](https://github.com/ZcashFoundation/zebra/blob/704f88eb412c9e5c0f0565c81250c6c50818275c/zebra-network/src/isolated/tor.rs#L88) (or whatever the latest version of `arti` uses). See the linked examples for how this works in `zebra-network`.
- The library should also support TCP connections. 

Tor onion addresses authenticate the server, but not the client. Tor did have client authentication in v2 onion services, but I'm not sure if it is implemented in v3. (v2 has been disabled.)

#### Authentication

Typically, proof of knowledge uses a challenge-response protocol, to avoid response re-use. A rough sketch of that protocol is:
- The server provides a random challenge value.
- The client responds with a hash of: a fixed prefix unique to this authentication method, the challenge value, and the secret (the password or the cookie file contents)
- The server checks that the hash matches the hash it generated from the challenge value and password

For proving knowledge of a key, we might just be able to use a prefixed hash of that key, because the connection is already authenticated and encrypted.

It's also possible to use client and server certificates, if we use TLS (or HTTPS).

</details>

## Next Steps

- [ ] #8153 
    - Finalise the evaluation criteria for the RPC library 
- [ ] #8154
    - Make a list of potential RPC libraries and evaluate them against the requirements
    - Choose a library 
- [ ] Create wrapper crate for using chosen library in Zebra
- [ ] Design and implement RPCs

**Comment by mpguerra** [comment]
Any other requirements that are not security requirements?

**Comment by mpguerra** [comment]
Some ideas for authenticating RPC connections were raised in #4575

**Comment by arya2** [comment]
Another consideration here could be ease-of-use, JSON is a bit easier to read during development.

**Comment by mpguerra** [comment]
@arya2 @upbqdn @oxarbitrage If there are no more requirements we can close this discussion as finished
